aiohttp_s3_client/credentials.py

```diff
@@ -0,0 +1,158 @@
+                os.getenv("AWS_SESSION_TOKEN", '')
+            ),
+            region=os.getenv('AWS_DEFAULT_REGION', region),
+            service=service,
+        )
+
+
+class ConfigCredentials(Credentials):
+    DEFAULT_CONFIG_PATH = Path.home() / '.aws' / 'credentials'
+
+    def __init__(
+        self, config_path: Union[str, Path] = DEFAULT_CONFIG_PATH, *,
+        region: str = '', service: str = "s3", profile: str = 'auto'
+    ):
+        credentials_file = os.getenv('AWS_SHARED_CREDENTIALS_FILE')
+        if credentials_file is not None:
+            config_path = Path(credentials_file)
+
+        if not isinstance(config_path, Path):
+            config_path = Path(config_path)
+
+        if not config_path.exists():
+            super().__init__(
+                access_key_id='', secret_access_key='',
+                region=region, service=service,
+            )
+            return
+
+        if profile == 'auto':
+            profile = os.getenv('AWS_PROFILE', 'default')
+
+        creds = configparser.ConfigParser()
+        creds.read(config_path)
+
+        if profile not in creds:
+            super().__init__(
+                access_key_id='', secret_access_key='',
+                region=region, service=service,
+            )
+            return
+
+        section = creds[profile]
+
+        super().__init__(
+            access_key_id=section.get('aws_access_key_id'),
+            secret_access_key=section.get('aws_secret_access_key'),
+            region=section.get('region'), service=service,
+        )
+
+
+class MetadataCredentials(Credentials):
+    METADATA_ADDRESS = '169.254.169.254'
+    METADATA_PORT = 80
+
+    def _fast_probe(self):
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+            return sock.connect_ex(
+                (self.METADATA_ADDRESS, self.METADATA_PORT)
+            ) == 0
+
+    def _fetch(self, suffix: str) -> bytes:
+        metadata_url = (
+            f'http://{self.METADATA_ADDRESS}:{self.METADATA_PORT}/latest'
+        )
+        with urlopen(f'{metadata_url}{suffix}') as f:
+            return f.read()
+
+    def _fetch_json(self, suffix: str) -> Any:
+        return json.loads(self._fetch(suffix))
+
+    def __init__(self, *, service: str = "s3", region: str = ""):
+        if not self._fast_probe():
+            super().__init__(
+                access_key_id='', secret_access_key='',
+                region=region, service=service,
+            )
+            return
+
+        region = self._fetch_json(
+            '/dynamic/instance-identity/document'
+        )['region']
+
+        role = self._fetch('/meta-data/iam/security-credentials').decode()
+        data = self._fetch_json(
+            f'/meta-data/iam/security-credentials/{role}'
+        )
+
+        super().__init__(
+            access_key_id=data.get('AccessKeyId'),
+            secret_access_key=data.get('SecretAccessKey') or data.get('Token'),
```

Comment:
as per aws docs, we need both secret access key and token for temporary credentials:

https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_use-resources.html

Reply:
Hmm. Token has short lifetime i guess. Is it means this should be fetch periodically?

Reply:
My hypothesis is that the response headers should contain the expiration time of the token. At least if I were designing this I would do so.

Reply:
they last 36hrs (no real docs) and they seem to serve fresh ones every 6 hrs.  certianly long term users of such a cred will have to take into account.   there is an "Expiration" field there in that json document as you predicted, but I'd say refreshing every six hours is probably "the right thing" to do. 

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html
https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/ec2-iam-roles.html
https://stackoverflow.com/questions/53609051/expiration-time-of-aws-ec2-instance-profile-credentials

maybe a def resfresh(self): pass in Credentials where in MetadataCredentials() it actually does the business.

Of course that would require more surgery of the aws-request-signer stuff.  

I hadn't considered this in my implementation.  and I'm not super fond of adding these sync urllib http calls in the code, in what could be a hot path.  of course, these are tiny calls.   in my usage, the clientsession is short lived on the order of minutes, which I think would be typical for 99% of users.

Reply:
Reworked please review it again. And I'm not still a AWS user @toppk do you want to test is in your sandbox?